src/core/models/payloads/transaction.c

```diff
@@ -164,26 +134,44 @@ size_t tx_essence_serialize(transaction_essence_t* es, byte_t buf[]) {
     memset(offset, 0, sizeof(uint32_t));
     offset += sizeof(uint32_t);
   }
-  return (offset - buf) / sizeof(byte_t);
+  return expected_bytes;
 }
 
-void tx_essence_free(transaction_essence_t* es) {
-  if (es) {
-    utxo_inputs_free(&es->inputs);
-    utxo_outputs_free(&es->outputs);
+transaction_essence_t* tx_essence_deserialize(byte_t buf[], size_t buf_len) {
+  if (buf == NULL || buf_len < 2) {
+    printf("[%s:%d] invalid paramters\n", __func__, __LINE__);
+    return NULL;
+  }
 
-    if (es->payload) {
-      // TODO support other payloads
-      indexation_free(es->payload);
-    }
-    free(es);
+  transaction_essence_t* es = tx_essence_new();
+
+  size_t offset = 0;
+
+  // transaction type
+  memcpy(&es->tx_type, &buf[offset], sizeof(uint8_t));
+  offset += sizeof(uint8_t);
+
+  es->inputs = utxo_inputs_deserialize(&buf[offset], buf_len - offset);
+  offset += utxo_inputs_serialize_len(es->inputs);
+
+  es->outputs = utxo_outputs_deserialize(&buf[offset], buf_len - offset);
```

Comment:
Check for NULL `es->outputs` is missing.

Reply:
Done.